rust/feature-flags/src/flag_matching.rs

```diff
@@ -1,75 +1,370 @@
 use crate::{
-    api::FlagError,
+    api::{FlagError, FlagValue, FlagsResponse},
     database::Client as DatabaseClient,
-    flag_definitions::{FeatureFlag, FlagGroupType, PropertyFilter},
+    flag_definitions::{FeatureFlag, FeatureFlagList, FlagGroupType, PropertyFilter},
     property_matching::match_property,
 };
+use anyhow::Result;
 use serde_json::Value;
 use sha1::{Digest, Sha1};
-use std::{collections::HashMap, fmt::Write, sync::Arc};
+use sqlx::FromRow;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Write;
+use std::sync::Arc;
+use tracing::error;
+
+type TeamId = i32;
+type DatabaseClientArc = Arc<dyn DatabaseClient + Send + Sync>;
+type GroupTypeIndex = i32;
 
 #[derive(Debug, PartialEq, Eq)]
 pub struct FeatureFlagMatch {
     pub matches: bool,
     pub variant: Option<String>,
-    //reason
-    //condition_index
-    //payload
 }
 
-#[derive(Debug, sqlx::FromRow)]
-pub struct Person {
-    pub properties: sqlx::types::Json<HashMap<String, Value>>,
+#[derive(Debug, FromRow)]
+pub struct GroupTypeMapping {
+    pub group_type: String,
+    pub group_type_index: GroupTypeIndex,
+}
+
+/// This struct is a cache for group type mappings, which are stored in a DB.  We use these mappings
+/// to look up group names based on the group aggregation indices stored on flag filters, which lets us
+/// perform group property matching.  We cache them per request so that we can perform multiple flag evaluations
+/// without needing to fetch the mappings from the DB each time.
+/// Typically, the mappings look like this:
+///
+/// let group_types = vec![
+///     ("project", 0),
+///     ("organization", 1),
+///     ("instance", 2),
+///     ("customer", 3),
+///     ("team", 4),  ];
+///
+/// But for backwards compatibility, we also support whatever mappings may lie in the table.
+/// These mappings are ingested via the plugin server.
+#[derive(Clone)]
+pub struct GroupTypeMappingCache {
+    team_id: TeamId,
+    failed_to_fetch_flags: bool,
+    group_types_to_indexes: Option<HashMap<String, GroupTypeIndex>>,
```

Comment:
Is there ever a case where these are validly `None` and you still might want this object to exist? (as opposed to just, say, empty hashmaps, for a team that has no group-types)? I'm wondering if you're better off having an `init(team_id, db_client)` function that constructs this object, hitting the DB to do so, and then you always _know_ your group-type map is in memory and usable (and that should simplify a lot of the code below)

Reply:
The only reason I might not want to seed the cache from the DB before doing any targeting in the case where we have properties that we're computing locally that don't require a DB hit – the goal for property override matching computations is that they don't hit the DB at all (other than for a flag definition if the flag wasn't already stored in Redis, which is handled separately), since we want these computations to be as fast as possible.  

That said, I think this feedback about making more explicit declarations of existence of these properties whenever I _do_ need them still makes sense.  For example, right now I've coupled the `FeatureFlagMatcher` with requiring a `GroupTypeMappingCache` on instantiation, but that might not actually be true; what I could do instead is start off with no cache and only create it whenever I'm in a block where I know I'm going to evaluate group properties... or do something different, I'm not sure exactly what the perfect improvement is here just yet, but I see your point.  Mostly I just wanted to explain that initializing the cache with the first DB hit isn't optimal, even if it makes reasoning about the code easier everywhere else.  But there's a solution here!